File: 1023-time-based-key-value-store/1023-time-based-key-value-store.go
type Pair struct {
    val string
    time int
}

type TimeMap struct {
    timeMap map[string][]*Pair
}


func Constructor() TimeMap {
    timeMap := make(map[string][]*Pair)

    return TimeMap{
        timeMap,
    }
}


func (this *TimeMap) Set(key string, value string, timestamp int)  {
    this.timeMap[key] = append(this.timeMap[key], &Pair{value, timestamp})
}


func (this *TimeMap) Get(key string, timestamp int) string {
    timeStamps := this.timeMap[key]

    if len(timeStamps) == 0 {
        return ""
    }

    left, right := 0, len(timeStamps) - 1

    for left <= right {
        mid := (left + right) >> 1

        if timeStamps[mid].time > timestamp {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    if right < 0 || timeStamps[right].time > timestamp {
        return ""
    }

    return timeStamps[right].val
}


/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */

File: 1023-time-based-key-value-store/1023-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.data = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.data[key].append((value, timestamp))

    def get(self, key: str, timestamp: int) -> str:
        pairs = self.data[key]

        if not pairs:
            return ""
        
        # timestamps = [pair[1] for pair in pairs]
        
        # idx = bisect_right(timestamps, timestamp)

        left, right = 0, len(pairs) - 1

        while left <= right:
            mid = (left + right) >> 1

            if pairs[mid][1] > timestamp:
                right = mid - 1
            else:
                left = mid + 1

        if right < 0:
            return ""
        
        return pairs[right][0]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)